frame/contracts/src/migration/v10.rs

```diff
@@ -0,0 +1,289 @@
+		trie_id: info.trie_id,
+		code_hash: info.code_hash,
+		storage_bytes: Default::default(),
+		storage_items: Default::default(),
+		storage_byte_deposit: Default::default(),
+		storage_item_deposit: Default::default(),
+		storage_base_deposit: Default::default(),
+	};
+	old::ContractInfoOf::<T>::insert(account, info);
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebugNoBound, TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(T))]
+pub struct DepositAccount<T: Config>(AccountIdOf<T>);
+
+impl<T: Config> Deref for DepositAccount<T> {
+	type Target = AccountIdOf<T>;
+
+	fn deref(&self) -> &Self::Target {
+		&self.0
+	}
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(T))]
+pub struct ContractInfo<T: Config> {
+	pub trie_id: TrieId,
+	deposit_account: DepositAccount<T>,
+	pub code_hash: CodeHash<T>,
+	storage_bytes: u32,
+	storage_items: u32,
+	pub storage_byte_deposit: BalanceOf<T>,
+	storage_item_deposit: BalanceOf<T>,
+	storage_base_deposit: BalanceOf<T>,
+}
+
+#[derive(Encode, Decode, MaxEncodedLen, DefaultNoBound)]
+pub struct Migration<T: Config> {
+	last_key: Option<BoundedVec<u8, ConstU32<256>>>,
+	_phantom: PhantomData<T>,
+}
+
+#[storage_alias]
+type ContractInfoOf<T: Config> =
+	StorageMap<Pallet<T>, Twox64Concat, <T as frame_system::Config>::AccountId, ContractInfo<T>>;
+
+impl<T: Config> Migrate for Migration<T> {
+	const VERSION: u16 = 10;
+
+	fn max_step_weight() -> Weight {
+		T::WeightInfo::v10_migration_step()
+	}
+
+	fn step(&mut self) -> (IsFinished, Weight) {
+		let mut iter = if let Some(last_key) = self.last_key.take() {
+			old::ContractInfoOf::<T>::iter_from(last_key.to_vec())
+		} else {
+			old::ContractInfoOf::<T>::iter()
+		};
+
+		if let Some((account, contract)) = iter.next() {
+			let min_balance = Pallet::<T>::min_balance();
+
+			// Store last key for next migration step
+			self.last_key = Some(iter.last_raw_key().to_vec().try_into().unwrap());
+
+			// Get the new deposit account address
+			let deposit_account: DepositAccount<T> =
+				DepositAccount(T::AddressGenerator::deposit_address(&account));
+
+			// Calculate the existing deposit, that should be reserved on the contract account
+			let old_deposit = contract
+				.storage_base_deposit
+				.saturating_add(contract.storage_item_deposit)
+				.saturating_add(contract.storage_byte_deposit);
+
+			// Unreserve the existing deposit
+			// Note we can't use repatriate_reserve, because it only works with existing accounts
+			let remaining = T::Currency::unreserve(&account, old_deposit);
+
+			if !remaining.is_zero() {
+				log::warn!(target: LOG_TARGET, "Partially unreserved. Remaining {:?} out of {:?} asked", remaining, old_deposit);
+			}
+
+			// Attempt to transfer the old deposit to the deposit account.
+			// We just want to leave the minimum balance on the contract account
+			let amount = old_deposit.saturating_sub(min_balance);
```

Comment:
Maybe I am missing something but this should allow you to unify the first to attempts.

```suggestion
			let amount = old_deposit.saturating_sub(min_balance).min(T::Currency::reducible_balance(&account, Preserve, Polite));
```

Reply:
ah yes good point